### ISSUE TITLE
User-facing diagnostic tools such as resolvectl should accept IDN

### ISSUE THREAD

**Issue by samueltardieu** [systemd/systemd#10737]
**systemd version the issue has been seen with**

2.39

**Used distribution**

ArchLinux.

**Expected behaviour you didn't see / Unexpected behaviour you saw**

`resolvectl` does not translate internationalized domain names (while `dig` does).

```bash
$ resolvectl query -t soa مليسيا
مليسيا: resolve call failed: All attempts to contact name servers or networks failed
$ dig مليسيا soa +short
a.mynic.net.my. hostmaster.mynic.net.my. 2018074874 1800 1800 604800 86400
```

Note that using the IDNA-encoded domain name works:

```bash
$ resolvectl query -t soa xn--mgbx4cd0ab
xn--mgbx4cd0ab IN SOA a.mynic.net.my hostmaster.mynic.net.my 2018074874 1800 1800 604800 86400

-- Information acquired via protocol DNS in 252.1ms.
-- Data is authenticated: no
```

**Proposed resolution**

Accept IDN as input in user-facing tools such as `resolvectl` and IDNA-encode such domain names as needed.

**Comment by poettering** [comment]
Note that we actually do IDNA encoding, but only if you ask for hostnames, i.e. without the `-t` switch.

When you ask for raw RRs we do not do IDNA encoding

**Comment by TriMoon** [comment]
> Note that we actually do IDNA encoding, but only if you ask for hostnames, i.e. without the `-t` switch.
> 
> When you ask for raw RRs we do not do IDNA encoding

Then that's what needs changing because asking for specific RRs should be separate logic from the actual hostname those RRs are asked for.
iow: Always use IDN when the hostname provided needs it. :wink:

**Comment by poettering** [comment]
Fixed by #18606
